### ISSUE TITLE
Compilation failing for Datalog program

### ISSUE THREAD

**Issue by azreika** [souffle-lang/souffle#1499]
I'm working off of the `examples/euclid` test, except with every relation set to output and a different set of facts. I've attached the set of Datalog files referenced below here: [files.zip](https://github.com/souffle-lang/souffle/files/4807572/files.zip).

Compiling `souffle -D- euclid.dl -F facts -o hello` works fine, and the output is as expected.

I altered the program to add magic set rules to the relations `Mod` and `ToMod` in the file `magic_euclid_working.dl`. Compiling it works and produces the correct output (as does using the interpreter).

I then altered it further to add magic set rules to the `Simplify` relation too, in `magic_euclid_crashing.dl`. The produced C++ file now does NOT compile (because of incorrect variable names?). Produces errors like:

```
compiler error: cannot compile source file hello.cpp
g++ -std=c++17   -fwrapv  -DUSE_NCURSES  -O0 -g3 -DUSE_LIBZ  -DUSE_SQLITE  -DRAM_DOMAIN_SIZE=64 -fopenmp  -march=native   -o[hidden]/euclid/hello hello.cpp  -ldl -lpthread -lsqlite3 -lz -lncursesw -I[hidden]/.local/bin/../include
hello.cpp: In lambda function:
hello.cpp:5663:118: error: ‘env8’ was not declared in this scope; did you mean ‘env4’?
 5663 | if( (ramBitCast<RamDomain>(env4[0]) == ramBitCast<RamDomain>((ramBitCast<RamSigned>(env2[0]) / ramBitCast<RamSigned>(env8[2])))) && (ramBitCast<RamDomain>(env4[1]) == ramBitCast<RamDomain>((ramBitCast<RamSigned>(env2[1]) / ramBitCast<RamSigned>(env8[2]))))) {
      |                                                                                                                      ^~~~
      |                                                                                                                      env4
...
```
after running `souffle -D- magic_euclid_crashing.dl -F facts -o hello`. The interpreter also produces the wrong answer now (BUT, the compilation failure is more concerning - I could have just muddled up the magic set somewhere since the program is quite large, but could be related).

The expected output with the `facts` folder attached is:
```
Gcd
first	second	gcd
===============
[[2, 1], nil]	nil	[[1, 1], nil]
===============
```

**Comment by mmcgr** [comment]
Sounds like a problem with levels, and expecting variables to exist. Does the RAM output make sense?

**Comment by azreika** [comment]
This looks odd:
```
BEGIN_DEBUG "Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]) :- \n   magic_Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]),\n   ToSimplify([[numerator,denominator],tail]),\n   Simplify_bf(tail,newTail),\n   numerator != 0,\n   GcdInt(numerator,denominator,gcd).\nin file [hidden]/euclid/magic_euclid_crashing.dl [293:1-298:38]"
     QUERY
      FOR t0 IN magic_Simplify_bb
       IF (NOT (magic_Simplify_bb = ∅))
        IF (t4.1 = (t2.1/t9.2))
         IF (t4.0 = (t2.0/t9.2))
          FOR t5 IN @delta_ToSimplify
           IF (NOT (@delta_ToSimplify = ∅))
            FOR t8 IN Simplify_bf
             IF (NOT (Simplify_bf = ∅))
              FOR t9 IN GcdInt
               IF (NOT (GcdInt = ∅))
                IF (NOT (t2.0,t2.1,t9.2) ∈ @delta_GcdInt)
                 IF (NOT (t1.1,t3.1) ∈ @delta_Simplify_bf)
                  IF (NOT ([[t2.0,t2.1],t1.1],[[(t2.0/t9.2),(t2.1/t9.2)],t3.1]) ∈ Simplify_bb)
                   IF (t2.0 != number(0))
                    IF (t1.1 = t8.0)
                     IF (t1.1 = t6.1)
                      IF (t2.0 = t9.0)
                       IF (t2.0 = t7.0)
                        IF (t3.1 = t8.1)
                         IF (t2.1 = t9.1)
                          IF (t2.1 = t7.1)
                           PROJECT ([[t2.0,t2.1],t1.1], [[(t2.0/t9.2),(t2.1/t9.2)],t3.1]) INTO @new_Simplify_bb
    END_DEBUG
```
considering that `t1` up to `t7` are used but not defined, and `t9` is used before it is defined. Is this normal in a RAM program?

At the end, this snippet is eventually transformed into this after all the RAM transformations:
```
984 | BEGIN_DEBUG "Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]) :- \n   magic_Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]),\n   ToSimplify([[numerator,denominator],tail]),\n   Simplify_bf(tail,newTail),\n   numerator != 0,\n   GcdInt(numerator,denominator,gcd).\nin file [hidden]/euclid/magic_euclid_crashing.dl [293:1-298:38]"
985 | QUERY
986 | IF ((((NOT (ToSimplify = ∅)) AND (NOT (magic_Simplify_bb = ∅))) AND (NOT (Simplify_bf = ∅))) AND (NOT (@delta_GcdInt = ∅)))
987 |     FOR t0 IN magic_Simplify_bb
988 |         IF (t2.0 != number(0))
989 |             IF ((t4.1 = (t2.1/t8.2)) AND (t4.0 = (t2.0/t8.2)))
990 |                 FOR t5 IN ToSimplify
991 |                     IF (t1.1 = t6.1)
992 |                         IF (((t2.1 = t7.1) AND (t2.0 = t7.0)) AND (t1.1,t3.1) ∈ Simplify_bf)
993 |                             FOR t8 IN @delta_GcdInt ON INDEX t8.0 = t2.0 AND t8.1 = t2.1
994 |                                 IF (NOT ([[t2.0,t2.1],t1.1],[[(t2.0/t8.2),(t2.1/t8.2)],t3.1]) ∈ Simplify_bb)
995 |                                     PROJECT ([[t2.0,t2.1],t1.1], [[(t2.0/t8.2),(t2.1/t8.2)],t3.1]) INTO @new_Simplify_bb
996 | END_DEBUG
```

**Comment by mmcgr** [comment]
That program has many layers of wrong. I don't think it even makes sense to try to interpret it.

Have you tried the debug report? That will show more about the transformations that led to this madness.

**Comment by azreika** [comment]
The first is a snippet right after AST to RAM translation, and the clause before translation looks fine:
```
Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]) :- 
   magic_Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]),
   ToSimplify([[numerator,denominator],tail]),
   Simplify_bf(tail,newTail),
   numerator != 0,
   GcdInt(numerator,denominator,gcd).
```
Might be something wrong going on with the AST to RAM translator? Possibly with the way records are being handled?

**Comment by azreika** [comment]
Okay, so the error seems to be coming from this rule:

```
Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]) :-·
   magic_Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]),
   ToSimplify([[numerator,denominator],tail]),
   Simplify_bf(tail,newTail),
   numerator != 0,
   GcdInt(numerator,denominator,gcd).
```
The program will only compile if both divisions by `gcd` in the first body rule of the clause are removed:

```
Simplify_bb([[numerator,denominator],tail],[[(numerator/gcd),(denominator/gcd)],newTail]) :-·
   magic_Simplify_bb([[numerator,denominator],tail],[[(numerator),(denominator)],newTail]),
   ToSimplify([[numerator,denominator],tail]),
   Simplify_bf(tail,newTail),
   numerator != 0,
   GcdInt(numerator,denominator,gcd).
```
The divisions in the head atom don't seem to be affecting compilation.

Once this was changed, the correct output was even produced from the magic file, so it does seem to be related to the interpreter producing the incorrect result.

I'll try to come up with a minimal example from this.

**Comment by azreika** [comment]
This much much smaller program has the same problem!

```
.type BoxedNumber = [x:number]
.decl a(x:BoxedNumber)
a([2]).

.decl b(x:number)
b(X) :- a([X/2]), c(X).

.decl c(x:number)
c(2).

.output b()
```

Running `souffle -D- program.dl -o bye` produces the following error:
```
compiler error: cannot compile source file bye.cpp
g++ -std=c++17   -fwrapv  -DUSE_NCURSES  -O0 -g3 -DUSE_LIBZ  -DUSE_SQLITE  -DRAM_DOMAIN_SIZE=64 -fopenmp  -march=native   -o[hidden]/euclid/badcomp/bye bye.cpp  -ldl -lpthread -lsqlite3 -lz -lncursesw -I/home/abdul/.local/bin/../include
bye.cpp: In lambda function:
bye.cpp:245:85: error: ‘env2’ was not declared in this scope; did you mean ‘env1’?
  245 | if( (ramBitCast<RamDomain>(env1[0]) == ramBitCast<RamDomain>((ramBitCast<RamSigned>(env2[0]) / ramBitCast<RamSigned>(RamSigned(2)))))) {
      |                                                                                     ^~~~
      |                                                                                     env1
failed to compile C++ source <bye.cpp>
```
And, if we just run the interpreter, we get a segfault:
```
Segmentation violation signal in rule:
b(X) :- 
   a([(X/2)]),
   c(X).
in file [hidden]/badcomp/program.dl [6:1-6:24]
```

**Comment by azreika** [comment]
Updating `ResolveAliasesTransformer` to resolve complex terms nested in records might fix the problem. At the moment, it seems to only resolve terms that appear directly as arguments in atoms.

E.g.
```
a(X) :- b(X), c(X+1).
```
is currently transformed by `ResolveAliasesTransformer` to
```
a(X) :- b(X), b(tmp0), tmp0 = X+1.
```
while
```
a(X) :- b(X), c([X+1]).
```
is not.

**Comment by azreika** [comment]
Changing `ResolveAliasesTransformer` to do this fixed it for the simple examples I've tried, including the minimal example above. However, with the original euclid example, a new problem now arises with an ungrounded variable assertion failure. The variable does not seem to be ungrounded, however. Here is a small example derived after experimenting with the euclid program: 

```
.decl Raise(x:number, y:number, z:number, w:number)
.decl ToMod_ff(x:number, y:number)
.decl Leading(x:number, y:Rec, z:number)
.decl Multiply(x:Rec, y:number, z:number)
.type Rec = [x:number, y:number]
Raise(multiplicationResult,(firstDegree-secondDegree),0,multiplicationResult) :-
   ToMod_ff(first,second),
   Leading(first,[firstNumerator,firstDenominator],firstDegree),
   Leading(second,[secondNumerator,secondDenominator],secondDegree),
   firstDegree >= secondDegree,
   Multiply([ _tmp_0, _tmp_2],second,multiplicationResult),
    _tmp_0 = (-(firstNumerator)*secondDenominator),
    _tmp_1 = -(firstNumerator),
    _tmp_2 = (firstDenominator*secondNumerator).
.output Raise()
ToMod_ff(0,1).
Leading(1, [0,1], 2).
Multiply([0,1], 2, 3).
```
The AST semantic checker does NOT find a problem with the groundedness, but the assertion still breaks during AST translation. The produced Datalog code using `--show transformed-datalog` seems to show that the Datalog code remains unchanged.

```
.decl Raise(x:number, y:number, z:number, w:number)
.decl ToMod_ff(x:number, y:number)
.decl Leading(x:number, y:Rec, z:number)
.decl Multiply(x:Rec, y:number, z:number)
.type Rec = [x:number, y:number]
Raise(multiplicationResult,(firstDegree-secondDegree),0,multiplicationResult) :-
   ToMod_ff(first,second),
   Leading(first,[firstNumerator,firstDenominator],firstDegree),
   Leading(second,[secondNumerator,secondDenominator],secondDegree),
   firstDegree >= secondDegree,
   Multiply([ _tmp_0, _tmp_2],second,multiplicationResult),
    _tmp_0 = (-(firstNumerator)*secondDenominator),
    _tmp_1 = -(firstNumerator),
    _tmp_2 = (firstDenominator*secondNumerator).
.output Raise()
ToMod_ff(0,1).
Leading(1, [0,1], 2).
Multiply([0,1], 2, 3).
```

`_tmp_0_` is said to be ungrounded. Changing the `_tmp_0` constraint to `_tmp_0 = -firstNumerator`, `_tmp_0 = secondDenominator`, or even just `_tmp_0 = firstNumerator * secondNumerator` fixes the problem, though. Will try to come up with a more minimal example for this separate issue. Either it really is ungrounded, in which case the groundedness check should catch it, or the translator is failing some other way.

**Comment by azreika** [comment]
Here's a minimal example:

```
.type BoxedNumber = [x:number]

.decl A(x:number)
A(X) :-
    B(X, [Y]),
    Y = -X*1.

.decl B(x:number, y:BoxedNumber)
B(1, [-1]).

.output A()
```

Replacing `Y = -X*1` with just `Y = -X` fixes the problem and produces the correct output, so this is another bug that needs fixing alongside the original problem.

**Comment by azreika** [comment]
Problem is fixed if all top-level body record argument functors are extracted. Will push up a fix soon.

**Comment by azreika** [comment]
On top of this issue, there are also missing RAM unpack statements from the output, which made finding this problem a lot trickier. Adding that fix to the PR, as well.

**Comment by azreika** [comment]
Resolved by PR #1502.
